contrib/storage-ipfs/src/main/java/org/apache/drill/exec/store/ipfs/IPFSHelper.java

```diff
@@ -0,0 +1,326 @@
+    // Rule out any non-local addresses as they might be NAT-ed external
+    // addresses that are not always reachable from the inside.
+    // But is it safe to assume IPFS always listens on loopback and local addresses?
+    List<MultiAddress> myAddrs = ((List<String>) res.get("Addresses"))
+        .stream()
+        .map(MultiAddress::new)
+        .filter(addr -> {
+          try {
+            InetAddress inetAddress = InetAddress.getByName(addr.getHost());
+            return inetAddress.isSiteLocalAddress()
+                || inetAddress.isLinkLocalAddress()
+                || inetAddress.isLoopbackAddress();
+          } catch (UnknownHostException e) {
+            return false;
+          }
+        })
+        .collect(Collectors.toList());
+    this.myself = new IPFSPeer(this, myID, myAddrs);
+
+    return this.myself;
+  }
+
+  public Multihash resolve(String prefix, String path, boolean recursive) {
+    Map<String, String> result = timedFailure(
+        (args) -> clientCompat.resolve((String) args.get(0), (String) args.get(1), (boolean) args.get(2)),
+        ImmutableList.<Object>of(prefix, path, recursive),
+        timeouts.get(IPFSTimeOut.FIND_PEER_INFO)
+    );
+    if (!result.containsKey("Path")) {
+      return null;
+    }
+
+    // the path returned is of form /ipfs/Qma...
+    String hashString = result.get("Path").split("/")[2];
+    return Multihash.fromBase58(hashString);
+  }
+
+  @FunctionalInterface
+  public interface ThrowingFunction<T, R, E extends Exception>{
+    R apply(final T in) throws E;
+  }
+
+  @FunctionalInterface
+  public interface ThrowingSupplier<R, E extends Exception> {
+    R get() throws E;
+  }
+
+  /**
+   * Execute a time-critical operation op within time timeout. Causes the query to fail completely
+   * if the operation times out.
+   * @param op a Function that represents the operation to perform
+   * @param in the parameter for op
+   * @param timeout consider the execution has timed out after this amount of time in seconds
+   * @param <T> Input type
+   * @param <R> Return type
+   * @param <E> Type of checked exception op throws
+   * @return R the result of the operation
+   * @throws E when the function throws an E
+   */
+  public <T, R, E extends Exception> R timedFailure(ThrowingFunction<T, R, E> op, T in, int timeout) throws E {
+    Callable<R> task = () -> op.apply(in);
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  public <R, E extends Exception> R timedFailure(ThrowingSupplier<R, E> op, int timeout) throws E {
+    Callable<R> task = op::get;
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  private <R, E extends Exception> R timedFailure(Callable<R> task, int timeout, TimeUnit timeUnit) throws E {
+    Future<R> res = executorService.submit(task);
+    try {
+      return res.get(timeout, timeUnit);
+    } catch (ExecutionException e) {
+      throw (E) e.getCause();
```

Comment:
Can we throw a `UserException` here as well?

Reply:
I'd like to keep it that way. The caller could recover from the exception.